### PR TITLE
Administrator can be create project

### PR DESCRIPTION
**Pull request by DawidCwiek**
[add] Modal with Form to create new project:
- Form have full validation
[add] Model column:
- After action create in project controller create two columns
- Columns have specific order by table in Projects table
[fix] Fix Rspec test after:
- modify loyalty for login or no login user 
- add "data" in Json response
- change FactoryBot

### CODE REVIEW

File: app/javascript/packs/project_form_modal.jsx [hunk 1]
```diff
@@ -1,44 +1,143 @@
...
+		};
+	};
+
+		handleValidation(){
+        let project = this.state.data;
+        let errors = {};
+        let formIsValid = true;
+
+        //title
+        if(!project["title"]){
```
Comment (Vakrim) [inline app/javascript/packs/project_form_modal.jsx]:
```suggestion
        if(!project.title){
```

File: app/javascript/packs/project_form_modal.jsx [hunk 2]
```diff
@@ -1,44 +1,143 @@
...
+           formIsValid = false;
+           errors["title"] = "Cannot be empty";
+        }
+
+        if(typeof project["title"] !== "undefined"){
```
Comment (Vakrim) [inline app/javascript/packs/project_form_modal.jsx]: you can use simpler check here `if(project.title)` or just add `else` clause to previous `if`

File: app/javascript/packs/project_form_modal.jsx [hunk 3]
```diff
@@ -1,44 +1,143 @@
...
-	state = {
-	  title: "",
-		desc: "",
-		key: "",
-	};
+  constructor(props) {
+    super(props);
+    this.state = {
+      modal: false,
+			data: {
```
Comment (Vakrim) [inline app/javascript/packs/project_form_modal.jsx]: you have very inconsistent indents in your file. Consider converting all tabs to spaces and/or using some code formatter like prettier

Comment (DawidCwiek) [reply app/javascript/packs/project_form_modal.jsx]: we use prettier after merge all pr

File: app/models/project.rb [hunk 4]
```diff
@@ -2,9 +2,22 @@
...
+  has_many :columns
   has_and_belongs_to_many :users
 
   validates :title, presence: true
   validates :desc, presence: true, length: { maximum: 160 }
   validates :key, presence: true, uniqueness: true, length: { maximum: 3 }
+
+  def add_columns_to_columns_order(value)
+    self.columns_order.push(value)
+    self.save
```
Comment (hirokiraj) [inline app/models/project.rb]: both selfs are redundant

File: spec/controllers/task_controller_spec.rb [hunk 5]
```diff
@@ -31,11 +32,14 @@
...
-    let(:valid_attributes) { { task: attributes_for(:task, project_id: project.id) } }
-    let(:invalid_attributes) { { task: attributes_for(:task, title: nil, project_id: project.id) } }
+    let(:column) { create(:column, project_id: project.id) }
+    let(:valid_attributes) { { task: attributes_for(:task, project_id: project.id, column_id: column.id) } }
+    let(:invalid_attributes) { { task: attributes_for(:task, title: nil, project_id: project.id, column_id: column.id) } }
 
     context 'valid attributes' do
       subject { post :create, params: valid_attributes, format: :json }
+      #before {subject}
+      #it {byebug}
```
Comment (hirokiraj) [inline spec/controllers/task_controller_spec.rb]: remove unused code

File: app/models/column.rb [hunk 6]
```diff
@@ -0,0 +1,10 @@
+class Column < ApplicationRecord
+  after_create :add_to_columns_order
+  belongs_to :project
+  has_many :task
+
+  def add_to_columns_order
+    project = Project.find(self.project_id)
```
Comment (hirokiraj) [inline app/models/column.rb]: under `project` variable, there is already a project

File: app/models/project.rb [hunk 7]
```diff
@@ -2,9 +2,22 @@
...
 class Project < ApplicationRecord
   has_many :task
+  has_many :columns
   has_and_belongs_to_many :users
 
   validates :title, presence: true
   validates :desc, presence: true, length: { maximum: 160 }
   validates :key, presence: true, uniqueness: true, length: { maximum: 3 }
+
+  def add_columns_to_columns_order(value)
```
[the remark this answers is not in the document]
Comment (hirokiraj) [reply app/models/project.rb]: `value` is more like a `column_id`?

File: spec/support/controller_macros.rb [hunk 8]
```diff
@@ -0,0 +1,11 @@
+module ControllerMacros
+
+  def login_user
+    before(:each) do
+      @request.env["devise.mapping"] = Devise.mappings[:user]
+      user = FactoryBot.create(:user)
+      #user.confirm! # or set a confirmed_at inside the factory. Only necessary if you are using the "confirmable" module
```
Comment (hirokiraj) [inline spec/support/controller_macros.rb]: unused code, let's remove it

**Pull request merged**